src/hotspot/os/posix/os_posix.cpp

```diff
@@ -537,6 +537,28 @@ bool os::has_allocatable_memory_limit(julong* limit) {
     *limit = (julong)rlim.rlim_cur;
     result = true;
   }
+
+#if defined(INCLUDE_TSAN) && defined(AARCH64)
+  // Current TSAN memory mapping for 48bits aarch64, a large continous space could be allocated between
+  // kMidAppMemBeg  = 0x0aaaa00000000ull and kMidAppMemEnd  = 0x0aaaf00000000ull, which is only 20GB size.
+  // Take 16GB here for safer allocation.
+  const julong max_avail_vmspace = 1024ULL * 1024ULL * 1024ULL * 16ULL; // 16GB
+  const u8 msb_in_aarch64 = 47; //Only support 48-bits space now.
+
+  // Based on current TSAN aarch64 memory mapping, libjvm.so will be loaded to high end address space.
+  u8 vm_addr_u8 = reinterpret_cast<u8>(&__FUNCTION__);
+  // High address in 48bits user space is like 0x0000ffffxxxxxxxx.
```

Comment:
Is this based on kHiAppMemBeg = 0x0ffff00000000ull and kHiAppMemEnd = 0x1000000000000ull?
Mentioning these two constants like kMidAppMemBeg and kMidAppMemEnd above would be a more helpful comment.

Reply:
dynamic libraries will be loaded between kHiAppMemBeg kHiAppMemEnd . it makes hard to allocate large continuous space.

Reply:
oh, I see what you mean.